libs/platform/src/lib/components/form/combobox/commons/base-combobox.ts

```diff
@@ -0,0 +1,749 @@
+            event.stopPropagation();
+
+            this.showList(false);
+        } else if (!event.ctrlKey && !KeyUtil.isKey(event, this.nonOpeningKeys)) {
+            this.showList(true);
+        }
+    }
+
+    /** Handle dialog dismissing, closes popover and sets backup data. */
+    dialogDismiss(term: string): void {
+        this.inputText = term;
+        this.showList(false);
+    }
+
+    /** Handle dialog approval, closes popover and propagates data changes. */
+    dialogApprove(): void {
+        this.showList(false);
+    }
+
+    protected get ds(): ComboBoxDataSource<any> {
+        return (<ComboBoxDataSource<any>>this.dataSource);
+    }
+
+    /** @hidden */
+    private _setupKeyboardService(): void {
+        this._menuKeyboardService.itemClicked
+            .pipe(takeUntil(this._destroyed))
+            .subscribe((index) => this.handleOptionItem(index));
+        this._menuKeyboardService.focusEscapeBeforeList = () => this.searchInputElement.nativeElement.focus();
+    }
+
+    /** Method that picks other value moved from current one by offset, called only when combobox is closed */
+    private _chooseOtherItem(offset: number): void {
+        const activeValue: OptionItem = this._getSelectItemByValue(this.inputTextValue);
+        const index: number = this._suggestions.findIndex(value => value === activeValue);
+
+        if (this._suggestions[index + offset]) {
+            this.handleOptionItem(this._suggestions[index + offset]);
+        }
+    }
+
+    /** @hidden */
+    private _getSelectItemByValue(displayValue: string): OptionItem {
+        return this._suggestions.find(value => value.label === displayValue);
+    }
+
+    /** @hidden */
+    private _setUpMobileMode(): void {
+        this._dynamicComponentService.createDynamicComponent(
+            { listTemplate: this.listTemplate, controlTemplate: this.controlTemplate },
+            ComboboxMobileComponent,
+            { container: this._element },
+            { injector: Injector.create({ providers: [{ provide: COMBOBOX_COMPONENT, useValue: this }] }) }
+        );
+    }
+
+    /** @hidden */
+    private _initializeDataSource(ds: FdpComboBoxDataSource<any>): void {
+        this._suggestions = [];
+
+        if (isDataSource(this.dataSource)) {
+            (this.dataSource as ComboBoxDataSource<any>).close();
+
+            if (this._dsSubscription) {
+                this._dsSubscription.unsubscribe();
+                this._dsSubscription = null;
+            }
+        }
+        // Convert whatever comes in as DataSource so we can work with it identically
+        this._dataSource = this._openDataStream(ds);
+    }
+
+    /** @hidden */
+    private _openDataStream(ds: FdpComboBoxDataSource<any>): ComboBoxDataSource<any> {
+        const initDataSource = this._toDataStream(ds);
+
+        if (initDataSource === undefined) {
+            throw new Error(`[dataSource] source did not match an array, Observable, or DataSource`);
```

Comment:
Won't this error be common across Platform controls supporting this data source pattern? Maybe its worth moving this message to `consts`

Reply:
it will be in the future, once we will have more implementations. right now we have only one. So no need to refactor now